### DISCUSSION TITLE
疑问：手动用DPlatformWindowHandle给Qt无边框窗口添加DXCB效果只有有时才能生效

### DISCUSSION THREAD

**Discussion by RigoLigoRLC** [linuxdeepin/developer-center#3351] [Developer Discussion | 开发者讨论]
我在Deepin 20.2.1环境下，创建一个无边框（`Qt::FramelessWindowHint`）的`QMainWindow window`、创建一个`DPlatformWindowHandle p(&window, &window)`，试图通过在DTK“注册”该窗口达到给普通窗口添加DXCB效果（如圆角、边框缩放）。期间遇到几个问题：

如果在`window.show()`之前创建DPlatformWindowHandle，则无论如何只能呈现无阴影、无法缩放的状态；

如果在`window.show()`之后创建，则大部分时候，窗口都会呈现无阴影、无法缩放的状态：

![image](https://user-images.githubusercontent.com/49363666/123678042-62c9ce80-d878-11eb-8d40-cc9448a3c795.png)

有的时候，窗口会具有与未注册到DXCB时一样的阴影，但仍无法缩放：

![image](https://user-images.githubusercontent.com/49363666/123677932-462d9680-d878-11eb-8c76-1d3229bfdfa4.png)

极个别时候窗口具有正常的圆角和正常的边框缩放：

![image](https://user-images.githubusercontent.com/49363666/123678100-770dcb80-d878-11eb-8d0b-4418f78f1c93.png)

作为对比，不注册到DXCB时无边框窗口的样式：

![image](https://user-images.githubusercontent.com/49363666/123678224-a4f31000-d878-11eb-8165-8cd26b53bcdf.png)

如果这样做能够生效，那么肯定有正确之处，但是我不知道这些问题发生的原因。如果这样的做法不正确，那么如何做才可以让普通的无边框QMainWindow具有DXCB的效果？

**Comment by RigoLigoRLC** [comment]
注：在此例中我使用一个外部动态库进行与dtkwidget的DPlatformWindowHandle的操作，用QPluginLoader加载；但与链接到程序内效果并无不同。

**Comment by justforlxz** [comment]
唔，这个操作我还没有试过，骚等我试一下。

**Comment by justforlxz** [comment]
似乎你不需要设置setWindowFlag，不管他就可以了。

![image](https://user-images.githubusercontent.com/12298476/124208034-53b57b80-db19-11eb-9f8d-c8c6ae293e1e.png)

**Comment by justforlxz** [comment]
经过测试，成功给dtk的同学新增了bug。

**Comment by kegechen** [comment]
貌似是qt5.15才会出现这种情况。

**Comment by RigoLigoRLC** [comment]
> 貌似是qt5.15才会出现这种情况。

我电脑硬盘小，宿主机Manjaro是5.15，客户机Deepin 20.2.2也是5.15，没条件测试其他版本哈

**Comment by kegechen** [comment]
Qt::FramelessWindowHint 这个flag是容易造成某些情况下没效果，原因暂时不明（可能窗管的问题）。建议使用时去掉这个flag，因为使用DPlatformWindowHandle 本来就是无边框的。
